filters.im

```diff
@@ -1556,64 +1556,47 @@ i_rgbdiff_image(i_img *im1, i_img *im2) {
 
   out = i_sametype_chans(im1, xsize, ysize, outchans);
 
-  if (im1->bits == i_8_bits && im2->bits == i_8_bits) {
-    i_color *line1 = mymalloc(xsize * sizeof(*line1));
-    i_color *line2 = mymalloc(xsize * sizeof(*line1));
+#code im1->bits == i_8_bits && im2->bits == i_8_bits
+    IM_COLOR *line1 = mymalloc(xsize * sizeof(*line1));
+    IM_COLOR *line2 = mymalloc(xsize * sizeof(*line1));
     i_img_dim x, y;
     int ch;
+    IM_WORK_T opaque;
+#ifdef IM_EIGHT_BIT
+    opaque = 255;
+#else
+    opaque = 1.0;
+#endif
 
     for (y = 0; y < ysize; ++y) {
-      i_glin(im1, 0, xsize, y, line1);
-      i_glin(im2, 0, xsize, y, line2);
+      IM_GLIN(im1, 0, xsize, y, line1);
+      IM_GLIN(im2, 0, xsize, y, line2);
       if (outchans != diffchans) {
         /* give the output an alpha channel since it doesn't have one */
-        for (x = 0; x < xsize; ++x)
-          line2[x].channel[diffchans] = 255;
-      }
-      for (x = 0; x < xsize; ++x) {
-        for (ch = 0; ch < diffchans; ++ch) {
-          if (ch == outchans-1) {
-            line2[x].channel[ch] = 255; /* ignore alpha channel */
-          } else {
-            line2[x].channel[ch] = i_abs(line1[x].channel[ch] - line2[x].channel[ch]);
-          }
+        for (x = 0; x < xsize; ++x) {
+          line2[x].channel[diffchans] = opaque;
         }
       }
-      i_plin(out, 0, xsize, y, line2);
-    }
-    myfree(line1);
-    myfree(line2);
-  }
-  else {
-    i_fcolor *line1 = mymalloc(xsize * sizeof(*line1));
-    i_fcolor *line2 = mymalloc(xsize * sizeof(*line2));
-    i_img_dim x, y;
-    int ch;
-
-    for (y = 0; y < ysize; ++y) {
-      i_glinf(im1, 0, xsize, y, line1);
-      i_glinf(im2, 0, xsize, y, line2);
-      if (outchans != diffchans) {
-        /* give the output an alpha channel since it doesn't have one */
-        for (x = 0; x < xsize; ++x)
-          line2[x].channel[diffchans] = 1.0;
-      }
       for (x = 0; x < xsize; ++x) {
         for (ch = 0; ch < diffchans; ++ch) {
           if (ch == outchans-1) {
-            line2[x].channel[ch] = 1.0; /* ignore alpha channel */
+            line2[x].channel[ch] = opaque;
           } else {
+#ifdef IM_EIGHT_BIT
+            line2[x].channel[ch] = i_abs(line1[x].channel[ch] - line2[x].channel[ch]);
+#else
             if (line1[x].channel[ch] != line2[x].channel[ch]) {
               line2[x].channel[ch] = i_abs(line1[x].channel[ch] - line2[x].channel[ch]);
             }
+#endif
```

Comment:
i_abs() takes an integer parameter, which means the floating point version will truncate the value.  A change like:

```
@@ -139,6 +139,7 @@ sub byte_samples {
     s/\bIM_LIMIT\(/IM_LIMIT_8(/g;
     s/\bIM_RENDER_LINE\(/i_render_line(/g;
     s/\bIM_FILL_COMBINE_F\b/i_fill_combine_f/g;
+    s/\bIM_ABS\b/abs/g;
   }
   
   @lines;
@@ -168,6 +169,7 @@ sub double_samples {
     s/\bIM_LIMIT\(/IM_LIMIT_double(/g;
     s/\bIM_RENDER_LINE\(/i_render_linef(/g;
     s/\bIM_FILL_COMBINE_F\b/i_fill_combinef_f/g;
+    s/\bIM_ABS\b/fabs/g;
   }
 
   @lines;
@@ -281,6 +283,10 @@ object.
 
 =item *
 
+IM_ABS(sample) - calculate the absolute value of an IM_WORK_T value.
+
+=item *
+
 IM_SAMPLE_MAX - maximum value for a sample
 
 =item *
```

would give you an abs you could use without further fiddling.

The comparison here is also a problem, since it leaves the sample from the "other" image in the output buffer when they sample are the same between the images.

Reply:
Very nice, addressed in my commit 6acc3d77cb7d41a87c747a7a4e6147b0299557c4. This makes the test xt/x90cmpversion.t fail because it expects a version bump by a new git tag. I do not want to interfere with a new tag, so would leave that to you.

Reply:
And after that I pushed commit 0b6e9ff113ca6c0b9053d85f4e01471d22fdbe0a which eliminates the alpha channel in the output file.

Reply:
Thank you so much for your patience and help.